File: frontend/src/jobProfiles.js
export const MAX_PROFILES = 3;

export const EVIDENCE_CATEGORIES = ["technical", "domain", "languages"];

export const CATEGORY_LABELS = {
  technical: "Verktyg & teknik",
  domain: "Metod & domän",
  languages: "Språk",
};

function newId() {
  return crypto.randomUUID().slice(0, 12);
}

export function defaultProfileLabel(headline = "") {
  const text = String(headline || "").trim();
  return text || "Mitt jobbsök";
}

export function emptyProfile({ label = "Mitt jobbsök", active = true } = {}) {
  return {
    id: newId(),
    label,
    active,
    evidence: [],
  };
}

export function normalizeJobProfiles(raw, headline = "") {
  const list = Array.isArray(raw) ? raw.slice(0, MAX_PROFILES) : [];
  const profiles = list.map((item) => ({
    id: String(item.id || newId()),
    label: String(item.label || defaultProfileLabel(headline)).trim().slice(0, 120),
    active: Boolean(item.active),
    evidence: Array.isArray(item.evidence)
      ? item.evidence.map((row) => ({
          id: String(row.id || newId()),
          term: String(row.term || "").trim(),
          category: EVIDENCE_CATEGORIES.includes(row.category) ? row.category : "domain",
          source: {
            type: row.source?.type || "manual",
            index: typeof row.source?.index === "number" ? row.source.index : null,
            label: String(row.source?.label || "").trim(),
          },
          confirmed: row.confirmed !== false,
        }))
      : [],
  }));

  if (!profiles.length) {
    profiles.push(emptyProfile({ label: defaultProfileLabel(headline), active: true }));
  }

  const activeCount = profiles.filter((profile) => profile.active).length;
  if (activeCount !== 1) {
    profiles.forEach((profile, index) => {
      profile.active = index === 0;
    });
  }

  const seen = new Set();
  for (const profile of profiles) {
    profile.evidence = profile.evidence.filter((item) => {
      const key = item.term.toLowerCase();
      if (!item.term || seen.has(key)) return false;
      seen.add(key);
      return true;
    });
  }

  return profiles;
}

export function activeProfile(profiles) {
  return profiles.find((profile) => profile.active) || profiles[0];
}

export function confirmedEvidence(profile) {
  return (profile?.evidence ?? []).filter((item) => item.confirmed && item.term);
}

export function profileHasEvidence(profile) {
  return confirmedEvidence(profile).length > 0;
}

export function anyProfileHasEvidence(profiles) {
  return profiles.some((profile) => profileHasEvidence(profile));
}

export function setActiveProfile(profiles, profileId) {
  return profiles.map((profile) => ({
    ...profile,
    active: profile.id === profileId,
  }));
}

export function addProfile(profiles, label) {
  if (profiles.length >= MAX_PROFILES) return profiles;
  const inactive = profiles.map((profile) => ({ ...profile, active: false }));
  return [
    ...inactive,
    emptyProfile({ label: label || `Profil ${profiles.length + 1}`, active: true }),
  ];
}

export function updateProfileLabel(profiles, profileId, label) {
  return profiles.map((profile) =>
    profile.id === profileId ? { ...profile, label: label.trim().slice(0, 120) } : profile
  );
}

export function experienceSourceLabel(index, row) {
  const title = String(row?.title || "").trim();
  return title ? `Erfarenhet ${index + 1}: ${title}` : `Erfarenhet ${index + 1}`;
}

export function educationSourceLabel(index, row) {
  const label = String(row?.degree || row?.school || "").trim();
  return label ? `Utbildning ${index + 1}: ${label}` : `Utbildning ${index + 1}`;
}

export function addEvidence(profile, { term, category, source }) {
  const text = String(term || "").trim();
  if (!text) return profile;
  const lowered = text.toLowerCase();
  const evidence = [];
  let replaced = false;
  for (const item of profile.evidence) {
    if (item.term.toLowerCase() === lowered) {
      evidence.push({
        ...item,
        term: text,
        category,
        source,
        confirmed: true,
      });
      replaced = true;
    } else {
      evidence.push(item);
    }
  }
  if (!replaced) {
    evidence.push({
      id: newId(),
      term: text,
      category,
      source,
      confirmed: true,
    });
  }
  return { ...profile, evidence };
}

export function removeEvidence(profile, term) {
  const lowered = String(term || "").toLowerCase();
  return {
    ...profile,
    evidence: profile.evidence.filter((item) => item.term.toLowerCase() !== lowered),
  };
}

export function applyEvidenceToProfiles(profiles, profileId, updater) {
  return profiles.map((profile) =>
    profile.id === profileId ? updater(profile) : profile
  );
}

export function evidenceByLabel(profile, label) {
  return profile.evidence.filter((item) => item.source?.label === label);
}

export function evidenceForSource(profile, sourceKey) {
  const [type, indexRaw] = sourceKey.split(":");
  const index = indexRaw === undefined ? null : Number(indexRaw);
  return profile.evidence.filter((item) => {
    const source = item.source || {};
    if (source.type !== type) return false;
    if (type === "manual" || type === "cv_section") return source.type === type;
    return source.index === index;
  });
}

export function groupEvidenceBySource(profile) {
  const groups = new Map();
  for (const item of confirmedEvidence(profile)) {
    const source = item.source || {};
    let key = "manual";
    if (source.type === "experience" && typeof source.index === "number") {
      key = `experience:${source.index}`;
    } else if (source.type === "education" && typeof source.index === "number") {
      key = `education:${source.index}`;
    } else if (source.label) {
      key = source.label;
    }
    if (!groups.has(key)) groups.set(key, []);
    groups.get(key).push(item);
  }
  return groups;
}

export function countSuggestions(bySource) {
  if (!bySource) return 0;
  return Object.values(bySource).reduce((total, items) => total + (items?.length ?? 0), 0);
}

export function removeSuggestion(bySource, sourceKey, term) {
  if (!bySource?.[sourceKey]) return bySource;
  const lowered = term.toLowerCase();
  return {
    ...bySource,
    [sourceKey]: bySource[sourceKey].filter((item) => item.term.toLowerCase() !== lowered),
  };
}

File: backend/core/job_profiles.py
"""Evidence-backed job profiles — searchable lenses tied to CV rows."""

from __future__ import annotations

import uuid
from typing import Any

from .skill_canonical import canonical_skill_label
from .skill_groups import (
    EMPTY_SKILL_GROUPS,
    SKILL_GROUP_KEYS,
    flatten_skill_groups,
    normalize_skill_groups,
)

MAX_PROFILES = 3
SOURCE_TYPES = ("experience", "education", "manual")
EVIDENCE_CATEGORIES = SKILL_GROUP_KEYS


def _new_id() -> str:
    return uuid.uuid4().hex[:12]


def _default_label(headline: str = "") -> str:
    text = headline.strip()
    return text if text else "Mitt jobbsök"


def empty_profile(*, label: str = "Mitt jobbsök", active: bool = True) -> dict:
    return {
        "id": _new_id(),
        "label": label[:120],
        "active": active,
        "evidence": [],
    }


def _normalize_source(raw: Any) -> dict:
    if not isinstance(raw, dict):
        raise ValueError("Evidence source must be an object.")
    source_type = raw.get("type", "manual")
    if source_type not in SOURCE_TYPES:
        raise ValueError(f"Unknown evidence source type: {source_type}")
    index = raw.get("index")
    if source_type in {"experience", "education"}:
        if not isinstance(index, int) or index < 0:
            raise ValueError(
                "Experience/education evidence requires a non-negative index."
            )
    else:
        index = None
    label = str(raw.get("label") or "").strip()
    return {"type": source_type, "index": index, "label": label}


def _normalize_evidence_item(raw: Any) -> dict:
    if not isinstance(raw, dict):
        raise ValueError("Evidence must be an object.")
    term = canonical_skill_label(str(raw.get("term") or ""))
    if not term:
        raise ValueError("Evidence term is required.")
    category = raw.get("category", "domain")
    if category not in EVIDENCE_CATEGORIES:
        raise ValueError(f"Unknown evidence category: {category}")
    confirmed = bool(raw.get("confirmed", True))
    item_id = str(raw.get("id") or _new_id())
    return {
        "id": item_id,
        "term": term,
        "category": category,
        "source": _normalize_source(raw.get("source") or {"type": "manual"}),
        "confirmed": confirmed,
    }


def normalize_job_profiles(
    raw: Any,
    *,
    headline: str = "",
) -> list[dict]:
    """Return 0–3 normalized profiles; create a default when empty."""
    if raw is None:
        raw = []
    if not isinstance(raw, list):
        raise ValueError("job_profiles must be a list.")

    profiles: list[dict] = []
    for item in raw[:MAX_PROFILES]:
        if not isinstance(item, dict):
            raise ValueError("Each job profile must be an object.")
        label = str(item.get("label") or _default_label(headline)).strip()[:120]
        if not label:
            raise ValueError("Profile label is required.")
        evidence_raw = item.get("evidence", [])
        if not isinstance(evidence_raw, list):
            raise ValueError("Profile evidence must be a list.")
        seen_terms: set[str] = set()
        evidence: list[dict] = []
        for row in evidence_raw:
            normalized = _normalize_evidence_item(row)
            lowered = normalized["term"].lower()
            if lowered in seen_terms:
                continue
            seen_terms.add(lowered)
            evidence.append(normalized)
        profiles.append(
            {
                "id": str(item.get("id") or _new_id()),
                "label": label,
                "active": bool(item.get("active")),
                "evidence": evidence,
            }
        )

    if not profiles:
        profiles = [empty_profile(label=_default_label(headline), active=True)]

    active_count = sum(1 for profile in profiles if profile["active"])
    if active_count != 1:
        for index, profile in enumerate(profiles):
            profile["active"] = index == 0
    return profiles


def active_profile(profiles: list[dict]) -> dict:
    for profile in profiles:
        if profile.get("active"):
            return profile
    return profiles[0]


def confirmed_evidence(profile: dict) -> list[dict]:
    return [item for item in profile.get("evidence", []) if item.get("confirmed")]


def evidence_to_skill_groups(evidence: list[dict]) -> dict[str, list[str]]:
    groups = {key: [] for key in SKILL_GROUP_KEYS}
    seen: set[str] = set()
    for item in evidence:
        if not item.get("confirmed"):
            continue
        term = item.get("term", "")
        lowered = term.lower()
        if not term or lowered in seen:
            continue
        seen.add(lowered)
        category = item.get("category", "domain")
        if category not in groups:
            category = "domain"
        groups[category].append(term)
    return normalize_skill_groups(groups)


def profiles_from_skill_groups(
    skill_groups: dict | None,
    *,
    headline: str = "",
) -> list[dict]:
    """Migrate legacy flat skill groups into one default profile."""
    groups = normalize_skill_groups(skill_groups or EMPTY_SKILL_GROUPS)
    evidence: list[dict] = []
    for category in SKILL_GROUP_KEYS:
        for term in groups.get(category, []):
            evidence.append(
                {
                    "id": _new_id(),
                    "term": term,
                    "category": category,
                    "source": {
                        "type": "manual",
                        "index": None,
                        "label": "Manuellt tillagd",
                    },
                    "confirmed": True,
                }
            )
    return [
        {
            "id": _new_id(),
            "label": _default_label(headline),
            "active": True,
            "evidence": evidence,
        }
    ]


def experience_source_label(index: int, row: dict) -> str:
    title = str(row.get("title") or "").strip()
    if title:
        return f"Erfarenhet {index + 1}: {title}"
    return f"Erfarenhet {index + 1}"


def education_source_label(index: int, row: dict) -> str:
    degree = str(row.get("degree") or row.get("school") or "").strip()
    if degree:
        return f"Utbildning {index + 1}: {degree}"
    return f"Utbildning {index + 1}"


def add_evidence_to_profile(
    profile: dict,
    *,
    term: str,
    category: str,
    source: dict,
    confirmed: bool = True,
) -> dict:
    """Return updated profile with one evidence item (deduped by term)."""
    label = canonical_skill_label(term)
    if not label:
        return profile
    lowered = label.lower()
    evidence = []
    replaced = False
    for item in profile.get("evidence", []):
        if item.get("term", "").lower() == lowered:
            evidence.append(
                {
                    **item,
                    "term": label,
                    "category": category,
                    "source": source,
                    "confirmed": confirmed,
                }
            )
            replaced = True
        else:
            evidence.append(item)
    if not replaced:
        evidence.append(
            {
                "id": _new_id(),
                "term": label,
                "category": category,
                "source": source,
                "confirmed": confirmed,
            }
        )
    return {**profile, "evidence": evidence}


def remove_evidence_term(profile: dict, term: str) -> dict:
    lowered = term.lower()
    evidence = [
        item
        for item in profile.get("evidence", [])
        if item.get("term", "").lower() != lowered
    ]
    return {**profile, "evidence": evidence}


def profile_skill_terms(profile: dict) -> list[str]:
    return flatten_skill_groups(evidence_to_skill_groups(confirmed_evidence(profile)))
